include/pybind11/detail/internals.h

```diff
@@ -82,10 +82,16 @@ struct type_equal_to {
 template <typename value_type>
 using type_map = std::unordered_map<std::type_index, value_type, type_hash, type_equal_to>;
 
+// Adapted from boost::hash_combine(). See also:
+// https://stackoverflow.com/a/27952689/7829525.
+inline void hash_combine(size_t& lhs, size_t rhs ) {
+    lhs ^= rhs + 0x9e3779b9 + (lhs << 6) + (lhs >> 2);
+}
+
 struct override_hash {
-    inline size_t operator()(const std::pair<const PyObject *, const char *>& v) const {
+    inline size_t operator()(const std::pair<const PyObject *, std::string>& v) const {
         size_t value = std::hash<const void *>()(v.first);
-        value ^= std::hash<const void *>()(v.second) + 0x9e3779b9 + (value<<6) + (value>>2);
+        hash_combine(value, std::hash<std::string>()(v.second));
         return value;
     }
 };
```

Comment:
Okay, this at least doesn't regress the hash like #2092 does. However this is still not what Boost's `hash_combine` does. Boost's equivalent would be:

```c++
size_t seed = 0;
size_t lhs = std::hash<const void*>()(v.first); // Same as here
size_t rhs = std::hash<std::string>()(v.second); // Same as here
seed ^= lhs + 0x9e3779b9 + (seed << 6) + (seed >> 2); // This line is missing!
seed ^= rhs + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```

Right now, instead of the first `seed ^= lhs...` we're just using `seed = lhs`. With initial seed being 0, boost's `hash_combine` would really do just `lhs += 0x9e3779b9;` before the `lhs ^= rhs + ...` line.

Full disclaimer, I do not claim to be an expert on good hashing functions. I'm just noticing the slight difference. I don't actually know if this should be addressed. I would however trust Boost's implementation on the grounds of "`boost::hash_combine` was written by someone who cares about hashes more than I do". Also, the `boost::hash_combine` has had this magic number since the beginning, [in 2005](https://www.boost.org/doc/libs/1_34_0/doc/html/boost/hash_combine.html).